### ISSUE TITLE
Avoid forming and after that returning lists in loops

### ISSUE THREAD

**Issue by PlugFox** [dart-code-checker/dart-code-metrics#443]
Avoid forming and after that returning lists in loops.
Because double memory consumption with more reduant code.

With iterable result you can transform it to Stream
For example, to avoid affecting the interface and event loops: https://dartpad.dev/6fad9242381bf89af4c1f6110c429e1c?null_safety=true

With iterable result you can much more effective with `.skip(n).toList()` or `.first` or `.isEmpty` afterwards.

Codegeneration, parsing, transforming must operate with Iterable not List, it's faster, lazy, less RAM consumption.

Example:
```dart
// BAD
List<int> bad() {
  final list = <int>[];
  for (var i = 0; i < 1000; i++) {
    list.add(i);
  }
  return list;
}

// GOOD
Iterable<int> good() sync* {
  for (var i = 0; i < 1000; i++) {
    yield i;
  }
}
Iterable<int> good() => Iterable<int>.generate(
      1000,
      (i) => i,
    );

```

**Comment by PlugFox** [comment]
As an example of incorrect code
![image](https://user-images.githubusercontent.com/7805176/131641188-6406170e-22e2-40e2-9beb-1c14f9a0efc5.png)

As an example of correct code
![image](https://user-images.githubusercontent.com/7805176/131641467-854fd8a2-201e-45f7-9098-35046572c438.png)

**Comment by PlugFox** [comment]
You can use Iterable something like this to avoid impact on Event loop and User Interface

**!!! THIS IS NOT TESTED AND PRODUCTION READY CODE !!!**
```dart
Stream<T> fixEventLoop<T extends Object?>(Iterable<T> iterable) {
  final stopwatch = Stopwatch();
  StreamSubscription<T>? sub;
  final sc = StreamController<T>(
    onListen: stopwatch.start,
    onPause: () => sub?.pause(),
    onResume: () => sub?.resume(),
    onCancel: () {
      sub?.cancel();
      stopwatch.stop();
    },
  );
  sub = Stream<T>.fromIterable(iterable).asyncMap((value) async {
    if (stopwatch.elapsedMilliseconds > 8) {
      await Future<void>.delayed(Duration.zero);
      stopwatch.reset();
    }
    return value;
  }).listen(
    sc.add,
    onDone: sc.close,
    onError: sc.addError,
    cancelOnError: true,
  );
  return sc.stream;
}
```

But with List it's doesn't make sense.

**Comment by incendial** [comment]
That's an interesting problem, but I think that the size of a list is a key issue here if we'll implement such rule - because for small lists, like from 5 to 100 items, it doesn't look like a need optimization and looks more like a false-positive to me. What do think? Maybe you have some measurements when the size of the list really starts matter?

**Comment by PlugFox** [comment]
@incendial if you need download and save some picture.
Of couse you can download binary data in Uint8List and after that save it as File.

But it would still be correct to connect the download stream with IOSink.

So it is here.
I thought you were doing these rules in order to write the correct code, and not protect the wrong one)

In addition, the "correct" option is shorter and more practical)
